stdlib/sql/from.go

```diff
@@ -106,76 +109,107 @@ func (s *FromSQLProcedureSpec) Copy() plan.ProcedureSpec {
+	readFn := func(ctx context.Context, rows *sql.Rows) (flux.Table, error) {
+		reader, err := newRowReader(rows)
+		if err != nil {
+			_ = rows.Close()
+			return nil, err
+		}
+		return read(ctx, reader, a.Allocator())
+	}
+	iterator := &sqlIterator{spec: spec, id: dsid, read: readFn}
+	return execute.CreateSourceFromIterator(iterator, dsid)
 }
 
-type SQLIterator struct {
-	id             execute.DatasetID
-	administration execute.Administration
-	spec           *FromSQLProcedureSpec
-	db             *sql.DB
-	reader         *execute.RowReader
-}
+var _ execute.SourceIterator = (*sqlIterator)(nil)
 
-var _ execute.SourceDecoder = (*SQLIterator)(nil)
+type sqlIterator struct {
+	spec  *FromSQLProcedureSpec
+	id    execute.DatasetID
+	read  func(ctx context.Context, rows *sql.Rows) (flux.Table, error)
+	table flux.Table
+	err   error
+	done  bool
+}
 
-func (c *SQLIterator) Connect(ctx context.Context) error {
+func (c *sqlIterator) connect(ctx context.Context) (*sql.DB, error) {
 	db, err := sql.Open(c.spec.DriverName, c.spec.DataSourceName)
-
 	if err != nil {
-		return err
+		return nil, err
 	}
-	if err = db.Ping(); err != nil {
-		return err
+	if err := db.Ping(); err != nil {
+		_ = db.Close()
+		return nil, err
 	}
-	c.db = db
-
-	return nil
+	return db, nil
 }
 
-func (c *SQLIterator) Fetch(ctx context.Context) (bool, error) {
-	rows, err := c.db.Query(c.spec.Query)
+func (c *sqlIterator) Next(ctx context.Context) bool {
+	if c.done {
+		return false
+	}
+	defer func() { c.done = true }()
+
+	// Connect to the database so we can execute the query.
+	db, err := c.connect(ctx)
 	if err != nil {
-		return false, err
+		c.err = err
+		return false
 	}
+	defer func() { _ = db.Close() }()
 
-	var reader execute.RowReader
-	switch c.spec.DriverName {
-	case "mysql":
-		reader, err = NewMySQLRowReader(rows)
-	case "postgres", "sqlmock":
-		reader, err = NewPostgresRowReader(rows)
-	default:
-		return false, fmt.Errorf("unsupported driver %s", c.spec.DriverName)
+	rows, err := db.QueryContext(ctx, c.spec.Query)
+	if err != nil {
+		c.err = err
+		return false
 	}
+	defer func() { _ = rows.Close() }()
 
+	table, err := c.read(ctx, rows)
 	if err != nil {
-		return false, err
+		c.err = err
+		return false
 	}
-	c.reader = &reader
+	c.table = table
+	return true
+}
 
-	return false, nil
+func (c *sqlIterator) Table() flux.Table {
+	return c.table
 }
 
-func (c *SQLIterator) Decode(ctx context.Context) (flux.Table, error) {
-	groupKey := execute.NewGroupKey(nil, nil)
-	builder := execute.NewColListTableBuilder(groupKey, c.administration.Allocator())
+func (c *sqlIterator) Release() {
+	// This iterator does not maintain any internal connections.
+}
 
-	reader := *c.reader
+func (c *sqlIterator) Err() error {
+	return c.err
+}
+
+// read will use the RowReader to construct a flux.Table.
+func read(ctx context.Context, reader execute.RowReader, alloc *memory.Allocator) (flux.Table, error) {
```

Comment:
I think we should promote this to a public, exported function, maybe UnGroupedRead(...)?

it'd be cool to have another helper function  GroupedRead()  that takes a function parameter that can compute a group key given a row from the row reader.  This can be done in a separate PR though.

Reply:
This was mostly meant as a helper function because the real thing I wanted was for the implementations of `RowReader` to just return tables, but they were already written and I didn't feel like refactoring multiple locations so I did this. It might be helpful to expose it though.

Reply:
yeah, it's more of a forward-thinking vision that once you implement the row reader interface, you can use this function for free.